core/src/main/java/com/linecorp/armeria/server/HttpHeaderUtil.java

```diff
@@ -0,0 +1,186 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.CharMatcher;
+import com.google.common.base.Splitter;
+import com.google.common.base.Splitter.MapSplitter;
+import com.google.common.base.Strings;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+
+import io.netty.util.AsciiString;
+import io.netty.util.NetUtil;
+
+/**
+ * A utility class which provides useful methods for handling HTTP headers.
+ */
+final class HttpHeaderUtil {
+    private static final Logger logger = LoggerFactory.getLogger(HttpHeaderUtil.class);
+
+    private static final Splitter CSV_SPLITTER = Splitter.on(',').omitEmptyStrings();
+    private static final MapSplitter TOKEN_SPLITTER =
+            Splitter.on(';').omitEmptyStrings().withKeyValueSeparator('=');
+    private static final CharMatcher QUOTED_STRING_TRIMMER = CharMatcher.is('"');
+
+    @VisibleForTesting
+    static final Function<String, String> FORWARDED_CONVERTER = value -> TOKEN_SPLITTER.split(value).get("for");
+
+    /**
+     * Returns an {@link InetAddress} of a client who initiates a request.
+     *
+     * @param headers the HTTP headers which is received from the client
+     * @param candidateHeaderNames a list of the HTTP header names that may contain a client address
+     * @param proxiedAddresses source and destination addresses retrieved from PROXY protocol header
+     * @param remoteAddress a remote endpoint of a channel
+     * @param filter the filter which evaluates an {@link InetAddress} can be used as a client address
+     * @see <a href="https://tools.ietf.org/html/rfc7239">Forwarded HTTP Extension</a>
+     * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For">X-Forwarded-For</a>
+     */
+    static InetAddress determineClientAddress(HttpHeaders headers,
+                                              List<AsciiString> candidateHeaderNames,
+                                              @Nullable ProxiedAddresses proxiedAddresses,
+                                              InetSocketAddress remoteAddress,
+                                              Predicate<InetAddress> filter) {
+        for (final AsciiString name : candidateHeaderNames) {
+            final InetAddress addr;
+            if (name.equals(HttpHeaderNames.FORWARDED)) {
```

Comment:
Perhaps we could check if the header value is null or empty before doing this comparison?